fossasia-gci-website.js

```diff
@@ -0,0 +1,209 @@
+                  break;
+              }
+            });
+          });
+        });
+    });
+  } else if(err.code !== 'ENOENT') {  // Something went wrong (excluding path not found)
+      console.log(err);
+    }
+  });
+}
+
+pathField.onclick = function(e) {
+  e.preventDefault();
+
+  // Set up dialog for opening repo's path
+  const options = {
+    title: 'Choose the directory where open/save the repo',
+    properties: [ 'openDirectory' ]
+  };
+
+  // Got the path from the dialog
+  ipcRenderer.on('openDialogReply', function(event, paths) {
+    if(paths !== null) {  // Check the dialog hasn't been cancelled
+      // Set the chosen path as pathField's value
+      pathField.value = paths[0];
+      // Update the statuses (changing the value won't trigger pathfield's oninput
+      // event)
+      updateStatuses();
+    }
+  });
+
+  // Display the dialog
+  ipcRenderer.send('openDialog', options);
+}
+
+pathField.oninput = function(e) {
+  e.preventDefault();
+
+  updateStatuses();
+}
+
+gciWebButton.onclick = function(e) {
+  e.preventDefault();
+
+  const path = pathField.value;
+
+  switch(gciWebButton.mode) {
+    // Either "upstream" or none of the remotes aren't configured
+    case 'remote':
+    case 'remote-upstream':
+      gciWebButton.disabled = true;
+      gciWebButton.style.cursor = 'not-allowed';
+      gciWebButton.innerHTML = 'Loading...';
+
+      document.querySelector('#gci-website__loading').className += ' is-active'; // Display the loading spinner
+
+      Git.Repository.open(path)
+      .then(function(repo) {
+        Git.Remote.create(repo, 'upstream', 'https://github.com/fossasia/gci15.fossasia.org.git');
+        gciWebButton.innerHTML = 'Done!';
+        updateStatuses();
+        document.querySelector('#gci-website__loading').className = 'mdl-spinner mdl-spinner--single-color mdl-js-spinner';  // Hide the loading spinner (by removing the 'is-active' class)
+        setTimeout(function() { // Keep the button disabled with the "Done!" message for a moment
+          gciWebButton.disabled = false;
+          gciWebButton.style.cursor = 'default';
+          gciWebButton.innerHTML = 'Clone';
+        }, 1500);
+      });
+      break;
+
+    // The remote "origin" isn't correctly configured
+    case 'remote-origin':
+      // TO IMPLEMENT WITH USER'S USERNAME, ONCE #9 IS MERGED
+      /*
+      Git.Repository.open(path)
+      .then(function(repo) {
+          Git.Remote.create(repo, 'upstream', customRemote);
+      });
+      */
+      break;
+
+    // The repo isn't cloned yet
+    default:
+      gciWebButton.disabled = true;
+      gciWebButton.style.cursor = 'not-allowed';
+      gciWebButton.innerHTML = 'Loading...';
+
+      document.querySelector('#gci-website__loading').className += ' is-active'; // Display the loading spinner
+
+      Git.Clone.clone('https://github.com/fossasia/gci15.fossasia.org.git', path, null)
+      .then(function(repo, err) {
```

Comment:
Javascript promises are a little different to callbacks. They are supposed to replace callbacks.

So the nodegit callback here does not work the same way with promises.

A promise compatible callback takes two parameters. The **success** object and the **err**. In this case **success** object is `repo`.

This is how you should use them in promises.
```javascript
Git.Clone.clone('https://github.com/fossasia/gci15.fossasia.org.git', path, null)
.then(function(repo) {
  // if operation succeeds we come here
  // repo stuff here
})
.catch(err){
  // if it fails we come here
  // error stuff
})
```

`then` and `catch` are special methods of the promise object returned from `Git.Clone.clone`. They will be called according to the success of the operation.

You can actually think of them as regular try catch block.